runtime/src/drivers.rs

```diff
@@ -416,4 +428,55 @@ impl Drivers {
         drivers.cert_chain = cert_chain;
         Ok(())
     }
+
+    pub fn pl_context_threshold_exceeded(
+        pl0_pauser: u32,
+        flags: u32,
+        locality: u32,
+        dpe: &DpeInstance,
+    ) -> CaliptraResult<()> {
+        let used_pl0_dpe_context_count = dpe
+            .count_contexts(|c: &Context| {
+                c.state != ContextState::Inactive && c.locality == pl0_pauser
+            })
+            .map_err(|_| CaliptraError::RUNTIME_INTERNAL)?;
+        // the number of used pl1 dpe contexts is the total number of used contexts
+        // minus the number of used pl0 contexts, since a context can only be activated
+        // from pl0 or from pl1. Here, used means an active or retired context.
+        let used_pl1_dpe_context_count = dpe
+            .count_contexts(|c: &Context| c.state != ContextState::Inactive)
+            .map_err(|_| CaliptraError::RUNTIME_INTERNAL)?
+            - used_pl0_dpe_context_count;
+        if Self::is_caller_pl1(pl0_pauser, flags, locality)
+            && used_pl1_dpe_context_count == PL1_DPE_ACTIVE_CONTEXT_THRESHOLD
+        {
+            return Err(CaliptraError::RUNTIME_PL1_USED_DPE_CONTEXT_THRESHOLD_EXCEEDED);
+        } else if !Self::is_caller_pl1(pl0_pauser, flags, locality)
+            && used_pl0_dpe_context_count == PL0_DPE_ACTIVE_CONTEXT_THRESHOLD
+        {
+            return Err(CaliptraError::RUNTIME_PL0_USED_DPE_CONTEXT_THRESHOLD_EXCEEDED);
+        }
+        Ok(())
+    }
+
+    pub fn is_caller_pl1(pl0_pauser: u32, flags: u32, locality: u32) -> bool {
+        flags & PL0_PAUSER_FLAG == 0 && locality != pl0_pauser
+    }
+
+    pub fn clear_tags_for_non_active_contexts(
+        dpe: &mut DpeInstance,
+        context_has_tag: &mut [U8Bool; MAX_HANDLES],
+        context_tags: &mut [u32; MAX_HANDLES],
+    ) {
+        (0..MAX_HANDLES).for_each(|i| {
+            if i < dpe.contexts.len()
+                && i < context_has_tag.len()
+                && i < context_tags.len()
+                && dpe.contexts[i].state != ContextState::Active
+            {
+                context_has_tag[i] = U8Bool::new(false);
+                context_tags[i] = 0;
+            }
+        });
+    }
```

Comment:
Similar comments for these, I think they might make sense as members for `Drivers`

Reply:
See other comment.